src/server/gc_worker/compaction_filter.rs

```diff
@@ -154,133 +216,294 @@ struct WriteCompactionFilter {
+        } else {
+            CompactionFilterDecision::Remove
+        };
+        Ok(decision)
+    }
+
+    // It's possible that elder versions than a MVCC-delete mark occurs in a higher level.
+    // So those versions needs to be handled by scan + delete.
+    fn handle_delete_mark(&mut self) {
+        if let Err(e) = self.handle_delete_mark_impl() {
+            error!(
+                "WriteCompactionFilter meets a fatal error and can't recover";
+                "err" => ?e,
+            );
+            // Because the section is executed in RocksDB compaction threads,
+            // `exit` seems better than `panic`.
+            unsafe { libc::exit(-1) };
+        }
+    }
+
+    fn handle_delete_mark_impl(&mut self) -> Result<(), String> {
+        let (mark, mut filtered) = match self.deleting_filtered.take() {
+            Some((x, y)) => (x, y.into_iter().peekable()),
+            None => return Ok(()),
+        };
+        debug_assert!(self.is_bottommost_level);
+
+        let mut valid = self.write_iter_seek_to(&mark)?;
         let mut write_iter = self.write_iter.take().unwrap();
         if valid && write_iter.key() == mark {
             // The delete mark itself is handled in `filter`.
-            valid = write_iter.next().unwrap();
+            valid = write_iter.next()?;
         }
 
         while valid {
             let (key, value) = (write_iter.key(), write_iter.value());
-            if truncate_ts(key) != self.key_prefix.as_slice() {
+            if truncate_ts(key)? != self.mvcc_key_prefix.as_slice() {
                 break;
             }
-            self.handle_filtered_write(parse_write(value));
-            self.delete_write_key(key);
-            valid = write_iter.next().unwrap();
+            let (_, current_ts) = split_ts(key)?;
+            let mut need_delete = true;
+            while let Some((ts, filtered_seqno)) = filtered.peek() {
+                match current_ts.cmp(ts) {
+                    CmpOrdering::Greater => break,
+                    CmpOrdering::Less => {
+                        let _ = filtered.next().unwrap();
+                    }
+                    CmpOrdering::Equal => {
+                        // It's possible that there are multiple `key_ts` in different levels.
+                        // A potential case is that a snapshot has been ingested.
+                        let seq = match write_iter.sequence() {
+                            Some(seqno) => seqno,
+                            // The iterator should support to get sequence numbers.
+                            None => return Err("Iterator can't get seqno".to_owned()),
+                        };
+                        debug_assert!(seq >= *filtered_seqno);
+                        // NOTE: in the bottommost level sequence could be 0. So it's required that
+                        // an ingested file's sequence is not 0 if it overlaps with the DB.
```

Comment:
What if the ingest file's sequence is zero, then need_delete would be false. But actually it should delete

Reply:
So the DB doesn't overlap with the ingested file.

Reply:
oh, got it. if need delete, the ingested file must be overlapped with db.